### PR TITLE
cli: Deflake TestRemoveDeadReplicas

### PR DESCRIPTION
**Pull request by bdarnell**
Review commit-by-commit; the last commit produces an ugly diff but it's just the removal of a for loop and dedenting everything by one level. ([whitespace-ignoring view of that commit](https://github.com/cockroachdb/cockroach/commit/160c81c83965bb1932a44153e7cdb0e10b5ce372?w=1))

### CODE REVIEW

**Review by darinpp — APPROVED** [review]
Looks good although I don't fully understand what part prevents a node for campaigning if it is the only surviving.

**Review by tbg — COMMENTED** [review]
> Reviewed 1 of 1 files at r1.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41736)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @bdarnell)

___
*[pkg/cli/debug_test.go, line 317 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41736#-LrhqkT46Thl6iyyuhAj:-LrhqkT46Thl6iyyuhAk:b-pdri0t) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4fe2a44428cb42b59f8ee6eb5ced0bb9c3650f8/pkg/cli/debug_test.go#L317)):*
> ```Go
> 					// It should not return VALID for a lease held by a node that
> 					// has been removed from the configuration.
> 					time.Sleep(10 * time.Second)
> ```

You're talking about the code below, right?

https://github.com/cockroachdb/cockroach/blob/7db032ed67003f38c7ddd7dcd2e6a8a4a98082b5/pkg/storage/replica_raft.go#L1445-L1456

Wouldn't that lease (owned by n1 and seen by n2) expire eventually (in <10s) and then n2 would campaign out of its own volition, avoiding the need to put an explicit Sleep?

I'd like to understand the failure better, and if the fix is straightforward, do the fix instead.

**Review by bdarnell — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/41736)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @tbg)

___
*[pkg/cli/debug_test.go, line 317 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/41736#-LrhqkT46Thl6iyyuhAj:-LritKzW9TdVt1WR5wMU:b-a4j484) ([raw file](https://github.com/cockroachdb/cockroach/blob/d4fe2a44428cb42b59f8ee6eb5ced0bb9c3650f8/pkg/cli/debug_test.go#L317)):*
<details><summary><i>Previously, tbg (Tobias Grieger) wrote…</i></summary><blockquote>

You're talking about the code below, right?

https://github.com/cockroachdb/cockroach/blob/7db032ed67003f38c7ddd7dcd2e6a8a4a98082b5/pkg/storage/replica_raft.go#L1445-L1456

Wouldn't that lease (owned by n1 and seen by n2) expire eventually (in <10s) and then n2 would campaign out of its own volition, avoiding the need to put an explicit Sleep?

I'd like to understand the failure better, and if the fix is straightforward, do the fix instead.
</blockquote></details>

`shouldCampaignOnWake` will transition from `false` to `true` when the old lease expires. But there's nothing that will cause a "wake" after that time. I didn't finish pulling on the thread but it looks like fixing this will require subtle changes to production code paths so even if we clean this up I probably wouldn't want to backport it, and I think the sleep is the better fix for now.

**Review by tbg — APPROVED** [review]

File: pkg/cli/debug_test.go [hunk 1]
```diff
@@ -145,254 +143,264 @@ func TestRemoveDeadReplicas(t *testing.T) {
...
+				//
+				// Set replication mode to manual so that TestCluster doesn't call
+				// WaitForFullReplication before we've decommissioned the nodes.
+				clusterArgs.ReplicationMode = base.ReplicationManual
+				clusterArgs.ParallelStart = true
+
+				// Sleep before restarting to allow all previous leases to expire.
+				// Without this sleep, the 2/4 case is flaky because sometimes
+				// node 1 will be left with an unexpired lease, which tricks
+				// some heuristics on node 2 into never campaigning even though it
```
Comment (tbg) [inline pkg/cli/debug_test.go]: Can you add the additional color from your comment in here?

**Comment by bdarnell** [comment]
CI is still red because this test is very slow in `stressrace`: it's only able to complete one of the two configurations in the 3m45s limit of the per PR stress run. There's nothing specific that takes a long time (aside from the 10s sleep), it's just that every little thing like replicating a range takes an appreciable amount of time. I'll try profiling to see where the time's going.

**Comment by bdarnell** [comment]
> it's just that every little thing like replicating a range takes an appreciable amount of time. I'll try profiling to see where the time's going.

No smoking guns here, things are just slow under the race detector. A single `testrace` run of TestRemoveDeadReplicas (both subtests) takes about 50 seconds of real time and 240 CPU seconds on my laptop. On each PR we run new/changed tests under stressrace with a timeout of 3m45s (= 225s), and one process per CPU, so we're just exceeding the limit of what can be done in a single test with our current configurations (albeit via a subtle and unintentional limit; I bet we have existing tests that wouldn't pass if they had to go through this "new/changed test" stressrace regime. And if I had changed more than one package at once, the time limit would be even lower since it would be subdivided between the packages)

Looks like we can either increase the limits (set in github-pull-request-make/main.go) or port this to a roachtest instead. I don't see a way to speed this up by enough given the race detector overhead.

**Comment by tbg** [comment]
I'd prefer updating the limits.

**Comment by knz** [comment]
FWIW I have disabled some `cockroach demo` tests under testrace for exactly the same reason, see #41044 and #41045.

**Pull request merged**